### PR TITLE
[SPARK-35396][SQL][TESTS][FOLLOWUP] Add a UT to check if a user-defined cachedBatch is completely released 

### PR DESCRIPTION
**Pull request by xuechendi**
### What changes were proposed in this pull request?
This PR is used to do add a UT to check if user-defined cached batch are completely released when clearCache called. 

### Why are the changes needed?
Add a new UT file RefCountedTestCachedBatchSerializerSuite.scala

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT is added, org.apache.spark.sql.execution.columnar.RefCountedTestCachedBatchSerializerSuite

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 1]
```diff
@@ -225,10 +231,31 @@ case class CachedRDDBuilder(
     _cachedColumnBuffers
   }
 
+  def manualClose[T <: CachedBatch](entry: T): T = {
+    val entryManualCloseTasks = Future {
+      entry match {
+        case o: AutoCloseable =>
+          try {
+            o.close
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:239]: We can't reuse the manualClose method defined in your previous PR?
I'm also not seeing that CachedBatch implementations have any resources to free - what are you looking at there?
In this case, maybe we don't want to even create the Future unless there is something AutoCloseable to close.

Comment (xuechendi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:239]: @srowen , the reason I also want to do manual close in InMemoryRelation is because since 3.1.1, this class supports user-defined Serializer through "spark.sql.cache.serializer" , and by this new config, user is able to create user-defined CachedBatch and store it inside InMemoryRelation. In our case, we want to define an Apache ArrowCachedBatch to store ColumnarBatch data inside memory or serialize to disk. 
That is why I am also thinking to add a manual close function here , and use AutoCloseable to make it more general.

The reason of adding future here, is because from my local test, releasing couple Gigabytes (ex:  10G) data synchronizely will still hold df.unpersist() for a while.
I also added an UT to explain my purpose with user-define CachedBatchSerializer and CachedBatch.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:239]: I'm not arguing against Future - just arguing that it's pointless to make a Future if there is nothing to close. The code could be rearranged (here and in the previous commit) to maybe take advantage of that? not sure. Also see comment in your other PR about a thread pool, perhaps.

Such an impl should be implementing finalize of course but I do agree it's OK to try to proactively close.

**Review by dongjoon-hyun — COMMENTED** [review]
Could you rebase to the master and revise this PR because we removed `Future` from the previous commit?

**Comment by xuechendi** [comment]
> Could you rebase to the master and revise this PR because we removed `Future` from the previous commit?

@dongjoon-hyun , yes, updated with latest commit

**Review by srowen — COMMENTED** [review]
This also seems fine to me. Wait, doesn't this need to change the previous implementation we merged that used Future?

**Comment by xuechendi** [comment]
> This also seems fine to me. Wait, doesn't this need to change the previous implementation we merged that used Future?

Previous implementation in MemoryStore is modified with @cloud-fan 's PR, https://github.com/apache/spark/pull/32743

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 2]
```diff
@@ -225,10 +228,25 @@ case class CachedRDDBuilder(
     _cachedColumnBuffers
   }
 
+  def manualClose[T <: CachedBatch](entry: T): T = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:231]: nit: `freeCachedBatch`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 3]
```diff
@@ -225,10 +228,25 @@ case class CachedRDDBuilder(
...
+    }
+    entry
+  }
+
   def clearCache(blocking: Boolean = false): Unit = {
     if (_cachedColumnBuffers != null) {
       synchronized {
         if (_cachedColumnBuffers != null) {
+          _cachedColumnBuffers.foreach(manualClose)
           _cachedColumnBuffers.unpersist(blocking)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:250]: Sorry I may miss some context. Doesn't `_cachedColumnBuffers.unpersist(blocking)` ask the memory store to free the memory entry and release the memory?

Comment (xuechendi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:250]: @cloud-fan , hmm, from our test, since _cachedColumnBuffers is RDD[CachedBatch], after _cachedColumnBuffers.unpersist(blocking) was called, the reference of all CachedBatch will be deleted, just like what happened in MemoryStore.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:250]: `RDD.unpersist` will go to the new code path you added in `MemoryStore`, and release the memory, or do I miss something?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:250]: Did you set `blocking` to true in your test?

Comment (xuechendi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:250]: > `RDD.unpersist` will go to the new code path you added in `MemoryStore`, and release the memory, or do I miss something?

Oh, I didn't realize that, I thought cachedBatch required an extra close if it hasn't been serialized/deserialized to persist... I just verified of what you said, and cachedBatch did get released without calling _cachedColumnBuffers.foreach(manualClose).

Thanks for pointing out! Anyway, in that case, do you think it will be helpful if I keep the UT to make sure user-defined cachedBatch is closed? or I will close this PR.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: It's always good to add more test cases!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 4]
```diff
@@ -204,7 +205,7 @@ case class CachedRDDBuilder(
     serializer: CachedBatchSerializer,
     storageLevel: StorageLevel,
     @transient cachedPlan: SparkPlan,
-    tableName: Option[String]) {
+    tableName: Option[String]) extends Logging {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:208]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala [hunk 5]
```diff
@@ -143,3 +143,109 @@ class CachedBatchSerializerSuite  extends QueryTest with SharedSparkSession {
...
+      cacheAttributes,
+      selectedAttributes,
+      conf)
+  }
+
+  override def supportsColumnarOutput(schema: StructType): Boolean = false
+  override def supportsColumnarInput(schema: Seq[Attribute]): Boolean = false
+}
+
+class RefCountedTestCachedBatchSerializerSuite extends QueryTest with SharedSparkSession {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala:215]: shall we put it in a new file?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala [hunk 6]
```diff
@@ -143,3 +143,109 @@ class CachedBatchSerializerSuite  extends QueryTest with SharedSparkSession {
...
+    val cached = df.cache()
+    // count triggers the caching action. It should not throw.
+    cached.count()
+
+    // Make sure, the DataFrame is indeed cached.
+    assert(spark.sharedState.cacheManager.lookupCachedData(cached).nonEmpty)
+    assert(DummyAllocator.getAllocatedMemory > 0)
+
+    // Drop the cache.
+    cached.unpersist()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala:246]: The default behavior is non-blocking. Shall we use `eventually` to wrap the final assert `assert(DummyAllocator.getAllocatedMemory == 0)`? Or do `cached.unpersist(blocking = true)` here?

Comment (xuechendi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala:246]: I see, may because this UT is only doing decrement, my observation is that it can go back to zero when it reached this line.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala (hunk 5 above)
Comment (xuechendi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala:215]: Ok, I'll move to a new file

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala:246]: The test can be flaky with async behavior.

Comment (xuechendi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/CachedBatchSerializerSuite.scala]: I see, changed to using blocking = true, and I verified locally, if I did a 1sec wait in close function, only blocking = true can pass the test.

**Review by cloud-fan — APPROVED** [review]
LGTM. Please update PR title and description.

**Comment by xuechendi** [comment]
> LGTM. Please update PR title and description.

Updated, thanks

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master for Apache Spark 3.2.0.
